jumpy/_custom_fns.py

```diff
@@ -0,0 +1,106 @@
+"""Custom implemented functions."""
+
+from __future__ import annotations
+
+from typing import Any, Callable, Sequence, TypeVar
+
+import numpy as onp
+
+from jumpy import is_jax_installed, ndarray
+from jumpy.core import is_jitted, which_dtype, which_np
+
+try:
+    import jax
+    import jax.numpy as jnp
+except ImportError:
+    jax, jnp = None, None
+
+
+F = TypeVar("F", bound=Callable)
+
+
+def array(object: Any, dtype=None) -> ndarray:
+    """Creates an array given a list."""
+    if dtype is None:
+        try:
+            np = which_np(*object)
+        except TypeError:
+            np = which_np(object)  # object is not iterable (e.g. primitive type)
+    else:
+        np = which_dtype(dtype)
+
+    return np.array(object, dtype)
+
+
+def index_update(x: ndarray, idx: ndarray, y: ndarray) -> ndarray:
+    """Pure equivalent of x[idx] = y."""
+    if which_np(x, idx, y) is jnp:
+        return jnp.array(x).at[idx].set(jnp.array(y))
+    else:
+        x = onp.copy(x)
+        x[idx] = y
+        return x
+
+
+def meshgrid(
+    *xi, copy: bool = True, sparse: bool = False, indexing: str = "xy"
+) -> ndarray:
+    """Create N-D coordinate matrices from 1D coordinate vectors."""
+    if which_np(xi[0]) is jnp:
+        return jnp.meshgrid(*xi, copy=copy, sparse=sparse, indexing=indexing)
+    else:
+        return onp.meshgrid(*xi, copy=copy, sparse=sparse, indexing=indexing)
+
+
+def take(tree: Any, i: ndarray | Sequence[int] | int, axis: int = 0) -> Any:
+    """Returns tree sliced by i."""
+    if not is_jax_installed:
+        raise NotImplementedError("This function requires the jax module")
+
+    np = which_np(i)
+    if isinstance(i, (list, tuple)):
+        i = np.array(i, dtype=int)
+    return jax.tree_util.tree_map(lambda x: np.take(x, i, axis=axis, mode="clip"), tree)
```

Comment:
This `take` is specifically meant for pytrees right? To be consistent, we should create a separate package `jumpy.tree_util` (similar to jax module structure) where you place pytree specific functions. In the future, we may place other pytree specific functions there as well? 

Also, this leaves us the option to implement an array only `jumpy.take` that follows the `numpy.take` and `jax.numpy.take` API (see [here](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.take.html#jax.numpy.take)) .

Reply:
Im not particularly sure why this function doesn't just run the respective `.take` function and needs to use the tree map function. I would be tempted to reduce complexity and simplify the function but there was probably a reason by the google for this change

Reply:
Brax uses a lot of tree maps, so I think they simply added it to jumpy for convenience. Furthermore, I would be surprised if people use `numpy.take` a lot, because in most cases you can do the same with basic indexing: `arr[i, j, k]`.

Nevertheless, I believe [this package](https://github.com/google/tree-math) already aims to make `jax.numpy` functions compatible with pytrees, so I agree that pytree-compatible numpy methods are outside of jumpy's scope.